lib/render/vp2RenderDelegate/basisCurves.cpp

```diff
@@ -0,0 +1,1700 @@
+                int i2 = std::min(line[2], maxIndex);
+                int i3 = std::min(line[3], maxIndex);
+
+                int v0 = curveIndices[i0];
+                int v1 = curveIndices[i1];
+                int v2 = curveIndices[i2];
+                int v3 = curveIndices[i3];
+
+                finalIndices[lineNum].Set(v0, v1, v2, v3);
+            }
+        }
+
+        return VtValue(finalIndices);
+    }
+
+    VtValue _BuildLinesIndexArray(const HdBasisCurvesTopology& topology)
+    {
+        std::vector<GfVec2i> indices;
+        VtArray<int> vertexCounts = topology.GetCurveVertexCounts();
+
+        int vertexIndex = 0;
+        int curveIndex = 0;
+        TF_FOR_ALL(itCounts, vertexCounts) {
+            for(int i = 0; i < *itCounts; i+= 2) {
+                indices.push_back(GfVec2i(vertexIndex, vertexIndex + 1));
+                vertexIndex += 2;
+            }
+            curveIndex++;
+        }
+
+        VtVec2iArray finalIndices(indices.size());
+        VtIntArray const &curveIndices = topology.GetCurveIndices();
+
+        // If have topology has indices set, map the generated indices
+        // with the given indices.
+        if (curveIndices.empty())
+        {
+            std::copy(indices.begin(), indices.end(), finalIndices.begin());
+        }
+        else
+        {
+            size_t lineCount = indices.size();
+            int maxIndex = curveIndices.size() - 1;
+
+            for (size_t lineNum = 0; lineNum < lineCount; ++lineNum)
+            {
+                const GfVec2i &line = indices[lineNum];
+
+                int i0 = std::min(line[0], maxIndex);
+                int i1 = std::min(line[1], maxIndex);
+
+                int v0 = curveIndices[i0];
+                int v1 = curveIndices[i1];
+
+                finalIndices[lineNum].Set(v0, v1);
+            }
+        }
+
+        return VtValue(finalIndices);
+    }
+
+    VtValue _BuildLineSegmentIndexArray(const HdBasisCurvesTopology& topology)
+    {
+        const TfToken basis = topology.GetCurveBasis();
+        const bool skipFirstAndLastSegs = (basis == HdTokens->catmullRom);
+
+        std::vector<GfVec2i> indices;
+        const VtArray<int> vertexCounts = topology.GetCurveVertexCounts();
+        bool wrap = topology.GetCurveWrap() == HdTokens->periodic;
+        int vertexIndex = 0; // Index of next vertex to emit
+        int curveIndex = 0;  // Index of next curve to emit
+        // For each curve
+        TF_FOR_ALL(itCounts, vertexCounts) {
+            int v0 = vertexIndex;
+            int v1;
+            // Store first vert index incase we are wrapping
+            const int firstVert = v0;
+            ++ vertexIndex;
+            for(int i = 1;i < *itCounts; ++i) {
+                v1 = vertexIndex;
+                ++ vertexIndex;
+                if (!skipFirstAndLastSegs || (i > 1 && i < (*itCounts)-1)) {
+                    indices.push_back(GfVec2i(v0, v1));
```

Comment:
```c++
indices.emplace_back(v0, v1);
```
Although if you can resize indices to the correct size up front, this code would be an order of magnitude faster than using emplace_back here.